genie-agent/src/main/java/com/netflix/genie/agent/execution/statemachine/JobExecutionStateMachineImpl.java

```diff
@@ -18,102 +18,266 @@
+    private AtomicReference<States> currentState = new AtomicReference<>(States.READY);
 
     /**
      * Constructor.
      *
-     * @param stateMachine     the state machine
-     * @param executionContext the execution context
-     * @param executionStages  the execution stages
+     * @param executionStages  the (ordered) list of execution stages
+     * @param executionContext the execution context passed across stages during execution
+     * @param listeners        the list of listeners
      */
     public JobExecutionStateMachineImpl(
-        final StateMachine<States, Events> stateMachine,
+        final List<ExecutionStage> executionStages,
         final ExecutionContext executionContext,
-        final List<ExecutionStage> executionStages
+        final Collection<JobExecutionListener> listeners
     ) {
-        this.stateMachine = stateMachine;
         this.executionStages = executionStages;
         this.executionContext = executionContext;
-        this.stateMachine
-            .getExtendedState()
-            .getVariables()
-            .put(ExecutionStage.EXECUTION_CONTEXT_CONTEXT_KEY, executionContext);
-        this.executionCompletionListener = new StateMachineListenerAdapter<States, Events>() {
-            @Override
-            public void stateMachineStopped(final StateMachine<States, Events> sm) {
-                log.debug("State machine execution complete");
-                completionLatch.countDown();
-            }
-        };
+
+        this.taskExecutor = new TaskExecutorBuilder()
+            .corePoolSize(1)
+            .maxPoolSize(1)
+            .threadNamePrefix("StateMachine")
+            .build();
+        this.taskExecutor.initialize();
+        this.listener = new CompositeListener(listeners);
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
     public void start() {
-        // Listen for state machine stop event
-        stateMachine.addStateListener(executionCompletionListener);
-        // Start the state machine execution
-        stateMachine.start();
-        // Kick off first transition
-        stateMachine.sendEvent(Events.START);
+
+        final boolean startedNow = runFuture.compareAndSet(
+            null,
+            this.taskExecutor.submit(this::run)
+        );
+
+        if (!startedNow) {
+            throw new IllegalStateException("State machine already running");
+        }
+    }
+
+    private void run() {
+        this.listener.stateMachineStarted();
+        for (final ExecutionStage executionStage : executionStages) {
+            final States state = executionStage.getState();
+            log.debug("Job execution stage: {} ({})", state, executionStage.getClass().getSimpleName());
+
+            this.currentState.set(state);
+            this.listener.stateEntered(state);
+
+            int retriesLeft = state.getTransitionRetries();
+            while (true) { // Not actually used, loop is broken out from inside.
```

Comment:
The contents of this loop are pretty large. Might be good to break out some of the sections into methods for readability/clarity

Reply:
Indeed. I am in the process of writing tests, once i have those i might move things around to make it more readable while still being sure i didn't break it.